crates/services/executor/src/executor.rs

```diff
@@ -916,12 +916,21 @@ where
                         if message.da_height() != da_height {
                             return Err(ExecutorError::RelayerGivesIncorrectMessages)
                         }
-                        block_storage_tx
-                            .storage_as_mut::<Messages>()
-                            .insert(message.nonce(), &message)?;
-                        execution_data
-                            .events
-                            .push(ExecutorEvent::MessageImported(message));
+                        let message_nonce = message.nonce();
+                        let message_event = if !block_storage_tx
+                            .storage_as_ref::<Messages>()
+                            .contains_key(message_nonce)?
+                        {
+                            block_storage_tx
+                                .storage_as_mut::<Messages>()
+                                .insert(message_nonce, &message)?;
+                            ExecutorEvent::MessageImported(message)
+                        } else {
+                            let reason =
+                                "Message with the same nonce already exists".to_string();
+                            ExecutorEvent::MessageIgnored { message, reason }
+                        };
```

Comment:
I'm not sure even how do we want to handle this case.

Right now you don't handle the situation mentioned by @Voxelot .

> we should probably just skip nonces that have already been seen before

The `Messages` table contains only the list of unspent messages. If a message was seen before but spent after, it can still be inserted into the state in this solution.

Reply:
>  If a message was seen before but spent after, it can still be inserted into the state in this solution.

You mean if a message with the same nonce gets spent before we check for nonce collisions, it will get added still. Right?

Reply:
Yes. And I think it is okay behavior if L1 contracts emit the message with the same nonce. It is not our business to decide is it bad or not in this situation=D

Reply:
Then maybe we want a data structure that allows us to store multiple messages with colliding nonces?

Reply:
No, only one message should be active at the time.